File: server/routes/components.ts
import { Router } from 'express';
import { v4 as uuid } from 'uuid';
import { prisma } from '../utils/prisma.js';
import { authGuard, requireRole } from '../middleware/auth.js';
import { writeAudit } from '../middleware/audit.js';

const router = Router();

function canReadComponent(user: { id: string; roles: string[] }, component: { ownerId: string }) {
  return user.roles.includes('ADMIN') || component.ownerId === user.id;
}

function canWriteComponent(user: { id: string; roles: string[] }, component: { ownerId: string }) {
  return user.roles.includes('ADMIN') || component.ownerId === user.id;
}

// GET / — list components
router.get('/', authGuard, async (req, res) => {
  const userId = req.user!.id;
  const isAdmin = req.user!.roles.includes('ADMIN');
  const where = isAdmin ? {} : { ownerId: userId };

  const rows = await prisma.component.findMany({
    where,
    orderBy: { updatedAt: 'desc' },
    include: {
      versions: {
        orderBy: { versionNo: 'desc' },
        take: 1,
      },
    },
  });

  res.json({ items: rows });
});

// GET /:id — get single component
router.get('/:id', authGuard, async (req, res) => {
  const { id } = req.params;
  const row = await prisma.component.findUnique({
    where: { id },
    include: {
      versions: {
        orderBy: { versionNo: 'desc' },
        take: 1,
      },
    },
  });

  if (!row) {
    res.status(404).json({ message: '元件不存在' });
    return;
  }
  if (!canReadComponent(req.user!, row)) {
    res.status(403).json({ message: '无权限查看该元件' });
    return;
  }

  res.json(row);
});

// POST / — create component
router.post('/', authGuard, requireRole('ADMIN', 'COMPONENT_EDITOR'), async (req, res) => {
  const { name, category, description } = req.body ?? {};
  if (!name || !category) {
    res.status(400).json({ message: 'name/category 不能为空' });
    return;
  }

  const created = await prisma.component.create({
    data: {
      name: String(name),
      category: String(category),
      description: description ? String(description) : '',
      ownerId: req.user!.id,
    },
  });

  await prisma.componentVersion.create({
    data: {
      componentId: created.id,
      versionNo: 1,
      snapshot: {
        schemaVersion: 1,
        shapeElements: [],
        pins: [],
        matrix: { connections: [] },
      },
      createdBy: req.user!.id,
    },
  });

  await writeAudit(req.user!.id, 'COMPONENT_CREATE', 'Component', created.id, {
    name: created.name,
    category: created.category,
  });

  res.status(201).json(created);
});

// PATCH /:id — update component
router.patch('/:id', authGuard, requireRole('ADMIN', 'COMPONENT_EDITOR'), async (req, res) => {
  const { id } = req.params;
  const { name, category, description } = req.body ?? {};

  const existing = await prisma.component.findUnique({ where: { id } });
  if (!existing) {
    res.status(404).json({ message: '元件不存在' });
    return;
  }
  if (!canWriteComponent(req.user!, existing)) {
    res.status(403).json({ message: '无权限修改该元件' });
    return;
  }

  const updates: Record<string, unknown> = {};
  if (name !== undefined) {
    const normalizedName = String(name).trim();
    if (!normalizedName) {
      res.status(400).json({ message: 'name 不能为空' });
      return;
    }
    updates.name = normalizedName;
  }
  if (category !== undefined) {
    const normalizedCategory = String(category).trim();
    if (!normalizedCategory) {
      res.status(400).json({ message: 'category 不能为空' });
      return;
    }
    updates.category = normalizedCategory;
  }
  if (description !== undefined) {
    updates.description = String(description);
  }
  if (Object.keys(updates).length === 0) {
    res.status(400).json({ message: '没有可更新字段' });
    return;
  }

  const updated = await prisma.component.update({
    where: { id },
    data: updates,
  });

  await writeAudit(req.user!.id, 'COMPONENT_UPDATE', 'Component', updated.id, {
    name: updates.name,
    category: updates.category,
  });

  res.json(updated);
});

// POST /:id/duplicate — duplicate component
router.post('/:id/duplicate', authGuard, requireRole('ADMIN', 'COMPONENT_EDITOR'), async (req, res) => {
  const { id } = req.params;
  const source = await prisma.component.findUnique({ where: { id } });

  if (!source) {
    res.status(404).json({ message: '元件不存在' });
    return;
  }

  if (!req.user!.roles.includes('ADMIN') && source.ownerId !== req.user!.id) {
    res.status(403).json({ message: '无权限复制该元件' });
    return;
  }

  const latestVersion = await prisma.componentVersion.findFirst({
    where: { componentId: source.id },
    orderBy: { versionNo: 'desc' },
  });

  const prefix = `${source.name}副本`;
  let newName = '';
  let next = 2;
  while (next <= 100) {
    const probe = await prisma.component.findFirst({ where: { name: `${prefix}${next}` } });
    if (!probe) {
      newName = `${prefix}${next}`;
      break;
    }
    next += 1;
  }
  if (!newName) {
    newName = `${prefix}${uuid().slice(0, 8)}`;
  }

  const created = await prisma.$transaction(async (tx) => {
    const comp = await tx.component.create({
      data: {
        name: newName,
        category: source.category,
        description: source.description,
        ownerId: req.user!.id,
      },
    });

    await tx.componentVersion.create({
      data: {
        componentId: comp.id,
        versionNo: 1,
        snapshot:
          latestVersion?.snapshot || {
            schemaVersion: 1,
            shapeElements: [],
            pins: [],
            matrix: { connections: [] },
          },
        createdBy: req.user!.id,
      },
    });

    return comp;
  });

  await writeAudit(req.user!.id, 'COMPONENT_DUPLICATE', 'Component', created.id, {
    sourceId: source.id,
    sourceName: source.name,
    newName,
  });

  res.status(201).json(created);
});

// DELETE /:id — delete component
router.delete('/:id', authGuard, requireRole('ADMIN', 'COMPONENT_EDITOR'), async (req, res) => {
  const { id } = req.params;
  const existing = await prisma.component.findUnique({ where: { id } });

  if (!existing) {
    res.status(404).json({ message: '元件不存在' });
    return;
  }
  if (!canWriteComponent(req.user!, existing)) {
    res.status(403).json({ message: '无权限删除该元件' });
    return;
  }

  await prisma.component.delete({ where: { id } });

  await writeAudit(req.user!.id, 'COMPONENT_DELETE', 'Component', id, {
    name: existing.name,
    category: existing.category,
  });

  res.status(204).send();
});

// GET /:id/versions — list component versions
router.get('/:id/versions', authGuard, async (req, res) => {
  const { id } = req.params;
  const component = await prisma.component.findUnique({ where: { id } });

  if (!component) {
    res.status(404).json({ message: '元件不存在' });
    return;
  }
  if (!canReadComponent(req.user!, component)) {
    res.status(403).json({ message: '无权限查看该元件版本' });
    return;
  }

  const versions = await prisma.componentVersion.findMany({
    where: { componentId: id },
    orderBy: { versionNo: 'desc' },
  });

  res.json({ items: versions });
});

// GET /:id/versions/:versionNo — get specific component version
router.get('/:id/versions/:versionNo', authGuard, async (req, res) => {
  const { id, versionNo: versionNoText } = req.params;
  const versionNo = Number(versionNoText);
  if (!Number.isInteger(versionNo) || versionNo <= 0) {
    res.status(400).json({ message: 'versionNo 不合法' });
    return;
  }

  const component = await prisma.component.findUnique({ where: { id } });
  if (!component) {
    res.status(404).json({ message: '元件不存在' });
    return;
  }
  if (!canReadComponent(req.user!, component)) {
    res.status(403).json({ message: '无权限查看该元件版本' });
    return;
  }

  const version = await prisma.componentVersion.findUnique({
    where: {
      componentId_versionNo: {
        componentId: id,
        versionNo,
      },
    },
  });
  if (!version) {
    res.status(404).json({ message: '版本不存在' });
    return;
  }

  res.json(version);
});

// POST /:id/versions — create component version
router.post('/:id/versions', authGuard, requireRole('ADMIN', 'COMPONENT_EDITOR'), async (req, res) => {
  const { id } = req.params;
  const { snapshot } = req.body ?? {};

  if (!snapshot || typeof snapshot !== 'object' || Array.isArray(snapshot)) {
    res.status(400).json({ message: 'snapshot 不能为空且必须为对象' });
    return;
  }

  const component = await prisma.component.findUnique({ where: { id } });
  if (!component) {
    res.status(404).json({ message: '元件不存在' });
    return;
  }
  if (!canWriteComponent(req.user!, component)) {
    res.status(403).json({ message: '无权限创建该元件版本' });
    return;
  }

  const created = await prisma.$transaction(async (tx) => {
    const latest = await tx.componentVersion.findFirst({
      where: { componentId: id },
      orderBy: { versionNo: 'desc' },
      select: { versionNo: true },
    });
    const nextVersionNo = (latest?.versionNo || 0) + 1;

    return tx.componentVersion.create({
      data: {
        componentId: id,
        versionNo: nextVersionNo,
        snapshot,
        createdBy: req.user!.id,
      },
    });
  });

  await writeAudit(req.user!.id, 'COMPONENT_VERSION_CREATE', 'ComponentVersion', created.id, {
    componentId: id,
    versionNo: created.versionNo,
  });

  res.status(201).json(created);
});

export default router;

File: src-tauri/src/commands/components.rs
use ecdraw_core::error::AppError;
use ecdraw_core::middleware;
use ecdraw_core::models::{Component, ComponentVersion};
use ecdraw_core::AppState;
use serde::{Deserialize, Serialize};
use serde_json::json;
use tauri::State;
use uuid::Uuid;

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ComponentWithVersion {
    #[serde(flatten)]
    pub component: Component,
    pub latest_version: Option<ComponentVersion>,
}

/// GET /api/components
#[tauri::command]
pub async fn list_components(
    state: State<'_, AppState>,
    token: String,
) -> Result<Vec<ComponentWithVersion>, AppError> {
    let claims = middleware::verify_auth(&token, &state.jwt_access_secret)?;

    let components = if claims.roles.contains(&"ADMIN".to_string()) {
        sqlx::query_as::<_, Component>(
            "SELECT * FROM components ORDER BY updated_at DESC"
        )
        .fetch_all(&state.pool)
        .await?
    } else {
        let user_id: Uuid = claims.sub.parse().map_err(|_| AppError::Auth("无效的用户标识".into()))?;
        sqlx::query_as::<_, Component>(
            "SELECT * FROM components WHERE owner_id = $1 ORDER BY updated_at DESC"
        )
        .bind(user_id)
        .fetch_all(&state.pool)
        .await?
    };

    let mut results = Vec::new();
    for c in components {
        let ver = sqlx::query_as::<_, ComponentVersion>(
            "SELECT * FROM component_versions WHERE component_id = $1 ORDER BY version_no DESC LIMIT 1"
        )
        .bind(c.id)
        .fetch_optional(&state.pool)
        .await?;
        results.push(ComponentWithVersion { component: c, latest_version: ver });
    }

    Ok(results)
}

/// GET /api/components/:id
#[tauri::command]
pub async fn get_component(
    state: State<'_, AppState>,
    token: String,
    id: String,
) -> Result<ComponentWithVersion, AppError> {
    let claims = middleware::verify_auth(&token, &state.jwt_access_secret)?;
    let cid: Uuid = id.parse().map_err(|_| AppError::BadRequest("无效的元件ID".into()))?;

    let c = sqlx::query_as::<_, Component>("SELECT * FROM components WHERE id = $1")
        .bind(cid)
        .fetch_optional(&state.pool)
        .await?
        .ok_or_else(|| AppError::NotFound("元件不存在".into()))?;

    let user_id: Uuid = claims.sub.parse().map_err(|_| AppError::Auth("无效的用户标识".into()))?;
    if !claims.roles.contains(&"ADMIN".to_string()) && c.owner_id != user_id {
        return Err(AppError::Forbidden("无权访问此元件".into()));
    }

    let ver = sqlx::query_as::<_, ComponentVersion>(
        "SELECT * FROM component_versions WHERE component_id = $1 ORDER BY version_no DESC LIMIT 1"
    )
    .bind(c.id)
    .fetch_optional(&state.pool)
    .await?;

    Ok(ComponentWithVersion { component: c, latest_version: ver })
}

/// POST /api/components
#[tauri::command]
pub async fn create_component(
    state: State<'_, AppState>,
    token: String,
    name: String,
    category: String,
    description: Option<String>,
) -> Result<Component, AppError> {
    let claims = middleware::verify_auth(&token, &state.jwt_access_secret)?;
    middleware::require_role(&claims, &["ADMIN", "COMPONENT_EDITOR", "DIAGRAM_EDITOR"])?;

    if name.is_empty() || category.is_empty() {
        return Err(AppError::BadRequest("名称和分类不能为空".into()));
    }

    let user_id: Uuid = claims.sub.parse().map_err(|_| AppError::Auth("无效的用户标识".into()))?;
    let component = sqlx::query_as::<_, Component>(
        "INSERT INTO components (name, category, description, owner_id) VALUES ($1, $2, $3, $4) RETURNING *"
    )
    .bind(&name)
    .bind(&category)
    .bind(&description)
    .bind(user_id)
    .fetch_one(&state.pool)
    .await?;

    let default_snapshot = json!({
        "schemaVersion": 1,
        "shapeElements": [],
        "pins": [],
        "matrix": { "connections": [] }
    });
    sqlx::query(
        "INSERT INTO component_versions (component_id, version_no, snapshot, created_by) VALUES ($1, 1, $2, $3)"
    )
    .bind(component.id)
    .bind(&default_snapshot)
    .bind(user_id)
    .execute(&state.pool)
    .await?;

    sqlx::query(
        "INSERT INTO audit_logs (user_id, action, target_type, target_id, payload) VALUES ($1, 'COMPONENT_CREATE', 'Component', $2, $3)"
    )
    .bind(user_id)
    .bind(component.id)
    .bind(json!({"name": &name, "category": &category}))
    .execute(&state.pool)
    .await?;

    Ok(component)
}

/// PATCH /api/components/:id
#[tauri::command]
pub async fn update_component(
    state: State<'_, AppState>,
    token: String,
    id: String,
    name: Option<String>,
    category: Option<String>,
    description: Option<String>,
) -> Result<Component, AppError> {
    let claims = middleware::verify_auth(&token, &state.jwt_access_secret)?;
    middleware::require_role(&claims, &["ADMIN", "COMPONENT_EDITOR", "DIAGRAM_EDITOR"])?;

    let cid: Uuid = id.parse().map_err(|_| AppError::BadRequest("无效的元件ID".into()))?;
    let c = sqlx::query_as::<_, Component>("SELECT * FROM components WHERE id = $1")
        .bind(cid)
        .fetch_optional(&state.pool)
        .await?
        .ok_or_else(|| AppError::NotFound("元件不存在".into()))?;

    let user_id: Uuid = claims.sub.parse().map_err(|_| AppError::Auth("无效的用户标识".into()))?;
    if !claims.roles.contains(&"ADMIN".to_string()) && c.owner_id != user_id {
        return Err(AppError::Forbidden("无权修改此元件".into()));
    }

    if name.is_none() && category.is_none() && description.is_none() {
        return Err(AppError::BadRequest("无更新内容".into()));
    }

    let name = name.unwrap_or(c.name);
    let category = category.unwrap_or(c.category);
    let description = description.or(c.description);

    let updated = sqlx::query_as::<_, Component>(
        "UPDATE components SET name = $1, category = $2, description = $3, updated_at = NOW() WHERE id = $4 RETURNING *"
    )
    .bind(&name)
    .bind(&category)
    .bind(&description)
    .bind(cid)
    .fetch_one(&state.pool)
    .await?;

    sqlx::query(
        "INSERT INTO audit_logs (user_id, action, target_type, target_id, payload) VALUES ($1, 'COMPONENT_UPDATE', 'Component', $2, $3)"
    )
    .bind(user_id)
    .bind(cid)
    .bind(json!({"name": &name, "category": &category}))
    .execute(&state.pool)
    .await?;

    Ok(updated)
}

/// DELETE /api/components/:id
#[tauri::command]
pub async fn delete_component(
    state: State<'_, AppState>,
    token: String,
    id: String,
) -> Result<(), AppError> {
    let claims = middleware::verify_auth(&token, &state.jwt_access_secret)?;
    middleware::require_role(&claims, &["ADMIN", "COMPONENT_EDITOR", "DIAGRAM_EDITOR"])?;

    let cid: Uuid = id.parse().map_err(|_| AppError::BadRequest("无效的元件ID".into()))?;
    let c = sqlx::query_as::<_, Component>("SELECT * FROM components WHERE id = $1")
        .bind(cid)
        .fetch_optional(&state.pool)
        .await?
        .ok_or_else(|| AppError::NotFound("元件不存在".into()))?;

    let user_id: Uuid = claims.sub.parse().map_err(|_| AppError::Auth("无效的用户标识".into()))?;
    if !claims.roles.contains(&"ADMIN".to_string()) && c.owner_id != user_id {
        return Err(AppError::Forbidden("无权删除此元件".into()));
    }

    sqlx::query("DELETE FROM diagram_instances WHERE component_id = $1")
        .bind(cid)
        .execute(&state.pool)
        .await?;

    sqlx::query("DELETE FROM components WHERE id = $1")
        .bind(cid)
        .execute(&state.pool)
        .await?;

    sqlx::query(
        "INSERT INTO audit_logs (user_id, action, target_type, target_id) VALUES ($1, 'COMPONENT_DELETE', 'Component', $2)"
    )
    .bind(user_id)
    .bind(cid)
    .execute(&state.pool)
    .await?;

    Ok(())
}

/// POST /api/components/:id/duplicate
#[tauri::command]
pub async fn duplicate_component(
    state: State<'_, AppState>,
    token: String,
    id: String,
) -> Result<Component, AppError> {
    let claims = middleware::verify_auth(&token, &state.jwt_access_secret)?;
    middleware::require_role(&claims, &["ADMIN", "COMPONENT_EDITOR", "DIAGRAM_EDITOR"])?;

    let cid: Uuid = id.parse().map_err(|_| AppError::BadRequest("无效的元件ID".into()))?;
    let source = sqlx::query_as::<_, Component>("SELECT * FROM components WHERE id = $1")
        .bind(cid)
        .fetch_optional(&state.pool)
        .await?
        .ok_or_else(|| AppError::NotFound("元件不存在".into()))?;

    let user_id: Uuid = claims.sub.parse().map_err(|_| AppError::Auth("无效的用户标识".into()))?;

    let latest_ver = sqlx::query_as::<_, ComponentVersion>(
        "SELECT * FROM component_versions WHERE component_id = $1 ORDER BY version_no DESC LIMIT 1"
    )
    .bind(cid)
    .fetch_optional(&state.pool)
    .await?;

    let mut new_name = format!("{}副本", source.name);
    let mut suffix = 2;
    loop {
        let count = sqlx::query_scalar::<_, i64>(
            "SELECT COUNT(*) FROM components WHERE name = $1"
        )
        .bind(&new_name)
        .fetch_one(&state.pool)
        .await?;
        if count == 0 { break; }
        new_name = format!("{}副本{}", source.name, suffix);
        suffix += 1;
        if suffix > 100 {
            new_name = format!("{}副本{}", source.name, Uuid::new_v4().to_string().split('-').next().unwrap());
            break;
        }
    }

    let mut tx = state.pool.begin().await?;
    let dup = sqlx::query_as::<_, Component>(
        "INSERT INTO components (name, category, description, owner_id) VALUES ($1, $2, $3, $4) RETURNING *"
    )
    .bind(&new_name)
    .bind(&source.category)
    .bind(&source.description)
    .bind(user_id)
    .fetch_one(&mut *tx)
    .await?;

    let snapshot = latest_ver.map(|v| v.snapshot).unwrap_or_else(|| json!({
        "schemaVersion": 1, "shapeElements": [], "pins": [], "matrix": { "connections": [] }
    }));
    sqlx::query(
        "INSERT INTO component_versions (component_id, version_no, snapshot, created_by) VALUES ($1, 1, $2, $3)"
    )
    .bind(dup.id)
    .bind(&snapshot)
    .bind(user_id)
    .execute(&mut *tx)
    .await?;

    tx.commit().await?;

    sqlx::query(
        "INSERT INTO audit_logs (user_id, action, target_type, target_id, payload) VALUES ($1, 'COMPONENT_DUPLICATE', 'Component', $2, $3)"
    )
    .bind(user_id)
    .bind(dup.id)
    .bind(json!({"source_id": cid, "name": &new_name}))
    .execute(&state.pool)
    .await?;

    Ok(dup)
}

/// GET /api/components/:id/versions
#[tauri::command]
pub async fn list_component_versions(
    state: State<'_, AppState>,
    token: String,
    id: String,
) -> Result<Vec<ComponentVersion>, AppError> {
    let claims = middleware::verify_auth(&token, &state.jwt_access_secret)?;
    let cid: Uuid = id.parse().map_err(|_| AppError::BadRequest("无效的元件ID".into()))?;

    let c = sqlx::query_as::<_, Component>("SELECT * FROM components WHERE id = $1")
        .bind(cid)
        .fetch_optional(&state.pool)
        .await?
        .ok_or_else(|| AppError::NotFound("元件不存在".into()))?;

    let user_id: Uuid = claims.sub.parse().map_err(|_| AppError::Auth("无效的用户标识".into()))?;
    if !claims.roles.contains(&"ADMIN".to_string()) && c.owner_id != user_id {
        return Err(AppError::Forbidden("无权访问此元件".into()));
    }

    let versions = sqlx::query_as::<_, ComponentVersion>(
        "SELECT * FROM component_versions WHERE component_id = $1 ORDER BY version_no DESC"
    )
    .bind(cid)
    .fetch_all(&state.pool)
    .await?;

    Ok(versions)
}

/// GET /api/components/:id/versions/:versionNo
#[tauri::command]
pub async fn get_component_version(
    state: State<'_, AppState>,
    token: String,
    id: String,
    version_no: i32,
) -> Result<ComponentVersion, AppError> {
    let claims = middleware::verify_auth(&token, &state.jwt_access_secret)?;
    let cid: Uuid = id.parse().map_err(|_| AppError::BadRequest("无效的元件ID".into()))?;

    let c = sqlx::query_as::<_, Component>("SELECT * FROM components WHERE id = $1")
        .bind(cid)
        .fetch_optional(&state.pool)
        .await?
        .ok_or_else(|| AppError::NotFound("元件不存在".into()))?;

    let user_id: Uuid = claims.sub.parse().map_err(|_| AppError::Auth("无效的用户标识".into()))?;
    if !claims.roles.contains(&"ADMIN".to_string()) && c.owner_id != user_id {
        return Err(AppError::Forbidden("无权访问此元件".into()));
    }

    if version_no < 1 {
        return Err(AppError::BadRequest("版本号必须为正整数".into()));
    }

    let version = sqlx::query_as::<_, ComponentVersion>(
        "SELECT * FROM component_versions WHERE component_id = $1 AND version_no = $2"
    )
    .bind(cid)
    .bind(version_no)
    .fetch_optional(&state.pool)
    .await?
    .ok_or_else(|| AppError::NotFound("版本不存在".into()))?;

    Ok(version)
}

/// POST /api/components/:id/versions
#[tauri::command]
pub async fn create_component_version(
    state: State<'_, AppState>,
    token: String,
    id: String,
    snapshot: serde_json::Value,
) -> Result<ComponentVersion, AppError> {
    let claims = middleware::verify_auth(&token, &state.jwt_access_secret)?;
    middleware::require_role(&claims, &["ADMIN", "COMPONENT_EDITOR", "DIAGRAM_EDITOR"])?;

    let cid: Uuid = id.parse().map_err(|_| AppError::BadRequest("无效的元件ID".into()))?;
    let c = sqlx::query_as::<_, Component>("SELECT * FROM components WHERE id = $1")
        .bind(cid)
        .fetch_optional(&state.pool)
        .await?
        .ok_or_else(|| AppError::NotFound("元件不存在".into()))?;

    let user_id: Uuid = claims.sub.parse().map_err(|_| AppError::Auth("无效的用户标识".into()))?;
    if !claims.roles.contains(&"ADMIN".to_string()) && c.owner_id != user_id {
        return Err(AppError::Forbidden("无权修改此元件".into()));
    }

    if !snapshot.is_object() || snapshot.is_array() {
        return Err(AppError::BadRequest("快照数据格式无效".into()));
    }

    let mut tx = state.pool.begin().await?;
    let latest_no = sqlx::query_scalar::<_, i32>(
        "SELECT COALESCE(MAX(version_no), 0) FROM component_versions WHERE component_id = $1"
    )
    .bind(cid)
    .fetch_one(&mut *tx)
    .await?;

    let version = sqlx::query_as::<_, ComponentVersion>(
        "INSERT INTO component_versions (component_id, version_no, snapshot, created_by) VALUES ($1, $2, $3, $4) RETURNING *"
    )
    .bind(cid)
    .bind(latest_no + 1)
    .bind(&snapshot)
    .bind(user_id)
    .fetch_one(&mut *tx)
    .await?;

    tx.commit().await?;

    sqlx::query(
        "INSERT INTO audit_logs (user_id, action, target_type, target_id, payload) VALUES ($1, 'COMPONENT_VERSION_CREATE', 'Component', $2, $3)"
    )
    .bind(user_id)
    .bind(cid)
    .bind(json!({"version_no": version.version_no}))
    .execute(&state.pool)
    .await?;

    Ok(version)
}
